### ISSUE TITLE
Use SQL Server datetime2 functions when translating DateTime.Now and DateTime.UtcNow

### ISSUE THREAD

**Issue by stevendarby** [dotnet/efcore#26780]
In the below code, `DateTime.Now` is translated to `GETDATE()` and `DateTime.UtcNow` is translated to `GETUTCDATE()` - these functions return SQL Server's `datetime`.

Should they use equivalent functions that return `datetime2` instead? These are `SYSDATETIME()` and `SYSUTCDATETIME()` respectively. The precision and range of `datetime2` is a closer (exact?) match to .NET `DateTime` so it feels like the more faithful translation, but I'm not sure if there are legacy reasons to use the older functions.

I can submit a PR to fix this if it's considered a valid proposal.

https://github.com/dotnet/efcore/blob/c8c6c21cafb9d25b7d8f3be316fd806cb2132de5/src/EFCore.SqlServer/Query/Internal/SqlServerDateTimeMemberTranslator.cs#L96-L110

**Comment by ajcvickers** [comment]
Note for triage: docs don't seem to indicate any downsides: https://docs.microsoft.com/en-us/sql/t-sql/functions/sysutcdatetime-transact-sql?view=sql-server-ver15

@ErikEJ Are you aware of any reason we shouldn't do this?

**Comment by roji** [comment]
Makes sense to me - especially since we map CLR DateTime to SQL Server `datetime2` by default, and not to `datetime`...

**Comment by ErikEJ** [comment]
No, seems the sensible thing to do - wonder why it was not done initially?

**Comment by ajcvickers** [comment]
Possibly not supported in some older versions?

**Comment by ErikEJ** [comment]
Yes, 2005 and earlier... 😄

**Comment by roji** [comment]
Note that IIRC there's implicit conversions between `datetime2` and `datetime`, so this wouldn't be breaking for code the expects `datetime`.

**Comment by stevendarby** [comment]
One thing I also wanted to note/question is the DateTimeOffset.UtcNow translation. It's currently `SYSUTCDATETIME`, which returns a `datetime2` (and is incidentally what DateTime.UtcNow should translate use).

When I do a query selecting DateTimeOffset.UtcNow into a DateTimeOffset property, the query does `CAST(SYSUTCDATETIME() AS datetimeoffset)`. I'm not entirely sure at which stage the cast is introduced and wondered if it is more correct to do an explicit convert as part of the translation?

**Comment by roji** [comment]
SQL Server seems to have an implicit cast from `datetime2` to `datetimeoffset`, resulting in a UTC `datetimeoffset`:

```sql
CREATE TABLE foo (dto datetimeoffset);
INSERT INTO foo VALUES (CAST('2020-01-01 12:00:00' AS datetime2));
SELECT * FROM foo; -- 2020-01-01 12:00:00.0000000 +00:00
```

As such, I don't think there's a reason to introduce an explicit cast - do you see an actual bug resulting from this?

**Comment by stevendarby** [comment]
@roji, no actual bug :) I don’t think I was clear: EF Core is already adding an explicit CAST, it’s just that I couldn’t see that in the actual translation code and concluded it must be happening at some other (mysterious) stage, and wondered if doing it at point of translation would be clearer. 

But it turns out it’s just because I didn’t link to the full set of code. Just under the lines I linked to is the actual return statement, which handles the casting when it’s datetimeoffset - d’oh!

**Comment by ajcvickers** [comment]
@stevendarby 

> I can submit a PR to fix this if it's considered a valid proposal.

We agreed in triage that this would be good, so a PR would be very welcome.

**Comment by roji** [comment]
See #36616 (dup) for a repro and some additional comments and doc links.

I think we should be able to translate to either GETDATE() or SYSGETDATE() based on what DateTime.Now is being compared too (similar to our regular type mapping inference logic - though this would be in the choice of what to translate too).

**Comment by stevendarby** [comment]
@roji, tbh I never quite got the point of trying to translate differently depending on what it's being compared to. Currently it always translates to a datetime even when compared to a datetime2 column, and SQL Server handles the implicit cast just fine. If it's changed to always translate to datetime2, the inverse casting will be handled just as well too.

The point for me was more that .NET DateTime maps much more closely to datetime2 and so should be translated to that.

**Comment by roji** [comment]
> Currently it always translates to a datetime even when compared to a datetime2 column, and SQL Server handles the implicit cast just fine. If it's changed to always translate to datetime2, the inverse casting will be handled just as well too.

That's true, but at least in theory that means that the representation of the current instant - to which we're comparing - isn't as precise as it could/should be (I agree it's unlikely to make a difference in many real-world apps, but it still seems a bit problematic). Just always translating to SYSGETDATE() and relying on implicit conversion from `datetime2` to `datetime` may indeed be sufficient, but it still feels like a bit of a smell to me...
